apps/homescreen/test/unit/app_test.js

```diff
@@ -1,36 +1,564 @@
+        appendChildSpy.restore();
+        done();
+      });
+    });
+
+    test('should insert a container in the right order', () => {
+      app.startupMetadata = [{ order: 1 }, { order: 2 }, { order: 3 }];
+      app.addIconContainer(0);
+      app.addIconContainer(2);
+
+      assert.equal(app.icons.children.length, 2);
+
+      app.addIconContainer(1);
+
+      assert.equal(app.icons.children.length, 3);
+      assert.equal(app.icons.children[0].order, 1);
+      assert.equal(app.icons.children[1].order, 2);
+      assert.equal(app.icons.children[2].order, 3);
+    });
+  });
+
+  suite('App#addAppIcon()', () => {
+    setup(() => {
+      stubCreateElement();
+    });
+
+    teardown(() => {
+      restoreCreateElement();
+    });
+
+    test('should call refresh() on a gaia-app-icon element', () => {
+      var refreshStub = sinon.stub(gaiaAppIconEl, 'refresh');
+      app.addAppIcon({ manifestURL: {} });
+      assert.isTrue(refreshStub.called);
+    });
+
+    test('should set entry points', () => {
+      app.addAppIcon({ manifestURL: {} }, { entryPoint1: {}, entryPoint2: {} });
+      assert.isNotNull(gaiaAppIconEl.entryPoint);
+    });
+
+    test('should be an app', () => {
+      app.addAppIcon({ manifestURL: {} });
+      assert.isNotNull(gaiaAppIconEl.app);
+      assert.isNull(gaiaAppIconEl.bookmark);
+    });
+
+    test('should be a bookmark', () => {
+      app.addAppIcon({});
+      assert.isNull(gaiaAppIconEl.app);
+      assert.isNotNull(gaiaAppIconEl.bookmark);
+    });
+  });
+
+  suite('App#storeAppOrder()', () => {
+    setup(() => {
+      app.metadata.mSetup();
+    });
+
+    test('should persist apps in sorted order', () => {
+      app.icons.appendChild(getIcon('abc'));
+      app.icons.appendChild(getIcon('def'));
+      app.icons.appendChild(getIcon('gih'));
+      app.storeAppOrder();
+      assert.deepEqual(app.metadata._data, [
+        { id: 'abc/', order: 0 },
+        { id: 'def/', order: 1 },
+        { id: 'gih/', order: 2 }
+      ]);
+    });
+  });
+
+  suite('App#snapScrollPosition()', () => {
+    test('should do nothing when there are no icons', () => {
+      app.snapScrollPosition(0);
+      assert.equal(app.scrollable.style.scrollSnapPointsY, '');
+    });
+
+    suite('with icons', () => {
+      setup(() => {
+        app.icons.appendChild(getIcon('abc'));
+        sinon.stub(app.icons.firstElementChild,
+          'getBoundingClientRect', () => {
+            return { height: 100 };
+          });
+        app.scrollable = {
+          clientHeight: 200,
+          style: {},
+          scrollTop: 0,
+          scrollTo: () => {
+          }
+        };
+      });
+
+      test('should snap to icon height when the screen is small', () => {
+        app.scrollable.clientHeight = 150;
+        app.snapScrollPosition(0);
+        assert.equal(app.scrollable.style.scrollSnapPointsY, 'repeat(100px)');
+      });
+
+      test('should snap to screen height if high enough', () => {
```

Comment:
This test isn't right, the code never snaps to screen height - it always snaps to as many icons as can fit whole in the screen (so if the clientHeight here was 390, it would be repeat(300px) and this test would fail).